### ISSUE TITLE
Compile instructions for bsdiff_jni.cc

### ISSUE THREAD

**Issue by PokeMMO** [andrewhayden/archive-patcher#123]
Would it be possible to get some more information/instructions on how to properly compile the native generator?

https://github.com/google/archive-patcher/blob/v2/generator/src/main/java/com/google/archivepatcher/generator/bsdiff/wrapper/bsdiff_jni.cc

**Comment by pspencil** [comment]
Yes. I have been trying to get that to work. It is a bit tricky since it
transitively depended on a lot of other projects, some of which are not in
grade/maven.

**Comment by PokeMMO** [comment]
I managed to get it to compile https://github.com/PokeMMO/archive-patcher/tree/native-lib-build

I've yet to verify it creates bit-for-bit equal results, but it runs.

**Comment by pspencil** [comment]
It doesn't create bit for bit equal results (but both are compatible).

If you wish to verify, you can try to enable the native integration test in
the grade conf under integration test directory.

**Comment by pspencil** [comment]
FYI, I am trying out the fix. 

Currently stuck at `make` complaining that `cannot find -lbrotlidec`

**Comment by PokeMMO** [comment]
Sorry, you also need to install a few packages via package manager.

For debian based apt:
```
libbz2-dev
libbrotli-dev
libdivsufsort-dev
```

**Comment by pspencil** [comment]
Hi, thanks. I figured it out. Btw, can you put a LICENSE file in the repo
you put above? You can copy from
https://github.com/google/archive-patcher/blob/master/LICENSE

**Comment by PokeMMO** [comment]
Was already in master, but added it to that branch too.

**Comment by pspencil** [comment]
Hi,

Thanks. I managed to have a script which builds the SO file. However, I am
struggling to integrate gradle with it. In particular:

1. Tests cannot automatically add the native library to load path (not sure
if this is feasible from gradle)
2. Uberjar (i.e., tools) cannot include the native lib.
3. Another concern I have is whether JNI will work when it is published to
maven.

I am not an expert in Gradle (so maybe it is capable). I am looking at
other options now like Maven.

**Comment by PokeMMO** [comment]
Maven will have the same issues as gradle. Several projects have done their own custom native loading implementations [libgdx](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/SharedLibraryLoader.java) and [LWJGL3](https://github.com/LWJGL/lwjgl3/blob/9bd6e6046c6098c1f905e3681f32b95c7ec3f532/modules/lwjgl/core/src/main/java/org/lwjgl/system/SharedLibraryLoader.java)

(Note the differences on archive mapping between libgdx and lwjgl3, libgdx uses a suffix while lwjgl3 now uses an android style arch folder)

These implementations load the .so from the classpath as a resource, extract it to a temporary location on disk, and then invoke System.load

[For my custom testing fork, I simply created a subproject with the .so as a jar resource.](https://github.com/PokeMMO/archive-patcher/commit/dcda8d056febd602667eed9430a6e2d5bc4528b6)

And before I want to use it, I call libgdx's `new SharedLibraryLoader().load("bsdiff");` (And if it fails, revert back to non-native flag)

Then in your tests, you can just have a `@BeforeClass` which loads the library.
